src/Features/Core/Portable/Formatting/FormattingCodeFixProvider.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Formatting
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic, Name = PredefinedCodeFixProviderNames.FixFormatting)]
+    [Shared]
+    internal class FormattingCodeFixProvider : SyntaxEditorBasedCodeFixProvider
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds
+            => ImmutableArray.Create(IDEDiagnosticIds.FormattingDiagnosticId);
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            context.RegisterCodeFix(
+                new MyCodeAction(c => FixOneAsync(context.Document, context.Diagnostics, c)),
+                context.Diagnostics);
+
+            return Task.CompletedTask;
+        }
+
+        protected async Task<Document> FixOneAsync(Document document, ImmutableArray<Diagnostic> diagnostics, CancellationToken cancellationToken)
+        {
+            var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var changes = new List<TextChange>();
+            foreach (var diagnostic in diagnostics)
+            {
+                if (!tree.Equals(diagnostic.Location.SourceTree)
+                    || !diagnostic.Properties.TryGetValue(FormattingDiagnosticAnalyzer.ReplaceTextKey, out var replacement))
+                {
+                    continue;
+                }
+
+                changes.Add(new TextChange(diagnostic.Location.SourceSpan, replacement));
+            }
+
+            changes.Sort((left, right) => left.Span.Start.CompareTo(right.Span.Start));
+
+            return document.WithText(text.WithChanges(changes));
+        }
+
+        protected override async Task FixAllAsync(Document document, ImmutableArray<Diagnostic> diagnostics, SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var updatedDocument = await Formatter.FormatAsync(document, options, cancellationToken).ConfigureAwait(false);
+            editor.ReplaceNode(editor.OriginalRoot, await updatedDocument.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false));
```

Comment:
This seems redundant as we are going from updatedDocument to updatedRoot and the [caller](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/CodeFixes/SyntaxEditorBasedCodeFixProvider.cs,46) uses the updatedRoot to again create an updatedDocument. Ideally, the abstract method `FixAllAsync` should return a task with an updated root or document, but that change might need to touch all code fixes overriding the API, which will be out of scope of this PR. Can we refactor `SyntaxEditorBasedCodeFixProvider` to have a virtual/abstract method that returns updated document, which you can override?

```
        protected async Task<Document> FixAllWithEditorAsync(
            Document document,
            Func<SyntaxEditor, Task> editAsync,
            CancellationToken cancellationToken)
        {
            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
            var editor = new SyntaxEditor(root, document.Project.Solution.Workspace);
 
            return await FixAllWithEditorAsync(document, diagnostics, editor, cancellationToken).ConfigureAwait(false);
        }

        protected virtual async Task<Document> FixAllWithEditorAsync(
            Document document,
            ImmutableArray<Diagnostic> diagnostics,
            SyntaxEditor editor,
            CancellationToken cancellationToken)
        {
            await editAsync(editor).ConfigureAwait(false);

            var newRoot = editor.GetChangedRoot();
            return document.WithSyntaxRoot(newRoot);
        }
```

Reply:
> Can we refactor SyntaxEditorBasedCodeFixProvider to have a virtual/abstract method that returns updated document, which you can override?

The NuGet package implementation of the code fix uses `DocumentBasedFixAllProvider`, which is simpler than `SyntaxEditorBasedCodeFixProvider` for these cases. We have the option of choosing either approach when we unify the implementations in the CodeStyle and EditorFeatures layers.